Mathlib/Algebra/Group/EvenFunction.lean

```diff
@@ -0,0 +1,132 @@
+lemma EvenFun.left_comp {g : α → β} (hg : EvenFun g) (f : β → γ) : EvenFun (f ∘ g) :=
+  (congr_arg f <| hg ·)
+
+/-- If `f` is even and `g` is odd, then `f ∘ g` is even. -/
+lemma EvenFun.comp_oddFun [Neg β] {f : β → γ} (hf : EvenFun f) {g : α → β} (hg : OddFun g) :
+    EvenFun (f ∘ g) :=
+  fun r ↦ by rw [comp_apply, hg, hf, comp_apply]
+
+/-- If `f` and `g` are odd, then `f ∘ g` is odd. -/
+lemma OddFun.comp_oddFun [Neg β] [Neg γ] {f : β → γ} (hf : OddFun f) {g : α → β} (hg : OddFun g) :
+    OddFun (f ∘ g) :=
+  fun r ↦ by rw [comp_apply, hg, hf, comp_apply]
+
+end composition
+
+lemma EvenFun.add [Add β] {f g : α → β} (hf : EvenFun f) (hg : EvenFun g) : EvenFun (f + g) := by
+  intro a
+  simp only [hf a, hg a, Pi.add_apply]
+
+lemma OddFun.add [SubtractionCommMonoid β] {f g : α → β} (hf : OddFun f) (hg : OddFun g) :
+    OddFun (f + g) := by
+  intro a
+  simp only [hf a, hg a, Pi.add_apply, neg_add]
+
+section smul
+
+variable {γ : Type*} {f : α → β} {g : α → γ}
+
+lemma EvenFun.smul_evenFun [SMul β γ] (hf : EvenFun f) (hg : EvenFun g) : EvenFun (f • g) := by
+  intro a
+  simp only [Pi.smul_apply', hf a, hg a]
+
+lemma EvenFun.smul_oddFun [Monoid β] [AddGroup γ] [DistribMulAction β γ]
+    (hf : EvenFun f) (hg : OddFun g) :
+    OddFun (f • g) := by
+  intro a
+  simp only [Pi.smul_apply', hf a, hg a, smul_neg]
+
+lemma OddFun.smul_evenFun [Ring β] [AddCommGroup γ] [Module β γ] (hf : OddFun f) (hg : EvenFun g) :
+    OddFun (f • g) := by
+  intro a
+  simp only [Pi.smul_apply', hf a, hg a, neg_smul]
+
+lemma OddFun.smul_oddFun [Ring β] [AddCommGroup γ] [Module β γ] (hf : OddFun f) (hg : OddFun g) :
+    EvenFun (f • g) := by
+  intro a
+  simp only [Pi.smul_apply', hf a, hg a, smul_neg, neg_smul, neg_neg]
+
+lemma EvenFun.const_smul [SMul β γ] (hg : EvenFun g) (r : β) : EvenFun (r • g) := by
+  intro a
+  simp only [Pi.smul_apply, hg a]
+
+lemma OddFun.const_smul [Monoid β] [AddGroup γ] [DistribMulAction β γ] (hg : OddFun g) (r : β) :
+    OddFun (r • g) := by
+  intro a
+  simp only [Pi.smul_apply, hg a, smul_neg]
+
+end smul
+
+section mul
+
+variable {R : Type*} [Mul R] {f g : α → R}
+
+lemma EvenFun.mul_evenFun (hf : EvenFun f) (hg : EvenFun g) : EvenFun (f * g) := by
+  intro a
+  simp only [Pi.mul_apply, hf a, hg a]
+
+lemma EvenFun.mul_oddFun [HasDistribNeg R] (hf : EvenFun f) (hg : OddFun g) : OddFun (f * g) := by
+  intro a
+  simp only [Pi.mul_apply, hf a, hg a, mul_neg]
+
+lemma OddFun.mul_evenFun [HasDistribNeg R] (hf : OddFun f) (hg : EvenFun g) : OddFun (f * g) := by
+  intro a
+  simp only [Pi.mul_apply, hf a, hg a, neg_mul]
+
+lemma OddFun.mul_oddFun [HasDistribNeg R] (hf : OddFun f) (hg : OddFun g) : EvenFun (f * g) := by
+  intro a
+  simp only [Pi.mul_apply, hf a, hg a, mul_neg, neg_mul, neg_neg]
+
+end mul
+
+/--
+If `f` is both even and odd, and its target is a torsion-free commutative additive group,
+then `f = 0`.
+-/
+lemma zero_of_evenFun_and_oddFun [AddCommGroup β] [NoZeroSMulDivisors ℕ β]
```

Comment:
These hypotheses seem way too strong to me, but I can't really think of a more granular typeclass that would cover them.

Reply:
Agreed. The "correct" assumption is that `β` be 2-torsion-free, which we can write as `IsSMulRegular β 2`, but that isn't a typeclass.